File: my-folder/0160-intersection-of-two-linked-lists/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func getIntersectionNode(headA, headB *ListNode) *ListNode {
    visited := make(map[*ListNode]bool)

    for curr := headA; curr != nil; curr = curr.Next {
        visited[curr] = true
    }

    for curr := headB; curr != nil; curr = curr.Next {
        if visited[curr] {
            return curr
        }
    }

    return nil
}

File: my-folder/0160-intersection-of-two-linked-lists/solution.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function getIntersectionNode(headA: ListNode | null, headB: ListNode | null): ListNode | null {
    if (headA == null || headB == null) {
        return null
    }

    let pA = headA
    let pB = headB

    while (pA != pB) {
        if (pA == null) {
            pA = headB
        } else {
            pA = pA.next
        }

        if (pB == null) {
            pB = headA
        } else {
            pB = pB.next
        }
    }

    return pA
};
